common/src/main/java/io/druid/collections/ReferenceCountingResourceHolder.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.collections;
+
+import com.metamx.common.ISE;
+import com.metamx.common.logger.Logger;
+
+import java.io.Closeable;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class ReferenceCountingResourceHolder<T> implements ResourceHolder<T>
+{
+  private static final Logger log = new Logger(ReferenceCountingResourceHolder.class);
+
+  private final Object lock = new Object();
+  private final T object;
+  private final Closeable closer;
+
+  private long refcount = 1;
```

Comment:
starting at 1 is pretty typical, it denotes the reference acquired by whoever newed up the object.